### DISCUSSION TITLE
Autocomplete with Cities sugestiom

### DISCUSSION THREAD

**Discussion by Jaaneek** [Budibase/budibase#7252] [Help]
Hi, how would you approach integrating Algolia/Google places or any other provider to autocomplete in budibase?

**Comment by andz-bb** [comment]
hey @Jaaneek - I don't see a way for you to have a field with an autocomplete that hits an api with every keystroke, although it's possible this could be implemented as a custom component in a future update.

If it helps at all, if you have an existing dataset of cities then it's possible to use that as a data source for an Options Picker component, and then you can type to filter the list and autocomplete the city. I built this out quickly using an existing csv dataset with cities and countries I found online. I imported this dataset from csv as two tables on my Budibase DB, one with all of the cities and their respective countries and one for each distinct country.

Then in my screen I added two Data Providers; one for countries and one for cities. At the highest level I added a Country Provider which I used as the datasource for my country field Options Picker. I then added a City Provider within the form I was using, and used the value from my country field as a filter for this data. This meant that I could have a City Options Picker showing all of the cities in whichever country I selected, and I could type into the field to further filter down the options. 

![country and city field](https://user-images.githubusercontent.com/110921612/184674353-9292951a-0d42-43bd-a557-426b0f0109fb.png)

I know it's not exactly what you are looking for with the 3rd party api integrations you mentioned but this solution seems to work pretty well in my playing around with it. If you need a hand with any part of it or would like it explained in more detail, let me know and I'd be happy to help 😊

**Reply by Jaaneek** [reply]
Hi,

Thanks for your input, I only need cities for my country, so I skipped the part with selecting countries.
But I have another problem, not all of the cities are being fetched.
I have a table with 2328 cities, but it only fetches 1899 of them. They are sorted by ID.
<img width="382" alt="image" src="https://user-images.githubusercontent.com/25470423/185238158-ada50e98-9f35-4606-92b2-1fa5229503e1.png">



Do you know if this behaviour is intentional?
Can you check if it works for you when you have more than 2000 records, if yes can you show a video with all the steps? 
Thank you!

<img width="228" alt="image" src="https://user-images.githubusercontent.com/25470423/185237651-363da9d1-833a-4ac8-98db-1f21e3338988.png">
<img width="246" alt="image" src="https://user-images.githubusercontent.com/25470423/185237724-4b8ae1de-7e97-4f74-8a79-8bdaee7d0f80.png">

**Reply by andz-bb** [reply]
Hey, that's a good spot, I had a look in the server code there and I'm seeing that for these full table queries there's a hard limit of 1000 records being returned, even when the limit you set on the Data Provider is higher. I might raise that as an issue here to see if we can limit and validate that field in the front end to be more transparent about what it would be returning on the response.

As a workaround would it be possible for you to do what I did with countries above but for the voivodeship of each city where you would select one first to filter the list of cities by? I'm not sure what that would be like from a UX perspective but it would be a way of reducing the number of cities to an amount that Budibase will support.
